lib/src/utf16.dart

```diff
@@ -7,21 +7,74 @@ import 'dart:typed_data';
 
 import 'package:ffi/ffi.dart';
 
-/// [Utf16] implements conversion between Dart strings and zero-terminated
-/// UTF-16 encoded "char*" strings in C.
+/// The contents of a native zero-terminated array of UTF-16 code units.
 ///
-/// [Utf16] is represented as a struct so that `Pointer<Utf16>` can be used in
-/// native function signatures.
+/// The Utf16 type itself has no functionality, it's only intended to be used
+/// through a `Pointer<Utf16>` representing the entire array. This pointer is
+/// the equivalent of a char pointer (`const wchar_t*`) in C code. The
+/// individual UTF-16 code units are stored in native byte order.
 class Utf16 extends Opaque {
-  /// Convert a [String] to a UTF-16 encoded zero-terminated C string.
+  /// Creates a zero-terminated [Utf16] code-unit array from [string].
   ///
-  /// If [string] contains NULL characters, the converted string will be truncated
-  /// prematurely. Unpaired surrogate code points in [string] will be preserved
-  /// in the UTF-16 encoded result. See [Utf16Encoder] for details on encoding.
+  /// If [string] contains NUL characters, the converted string will be truncated
+  /// prematurely.
   ///
-  /// Returns a [allocator]-allocated pointer to the result.
+  /// Returns an [allocator]-allocated pointer to the result.
+  @Deprecated('Use StringUtf16Pointer.toNativeUtf16 instead.')
   static Pointer<Utf16> toUtf16(String string, {Allocator allocator = calloc}) {
-    final units = string.codeUnits;
+    return string.toNativeUtf16(allocator: allocator);
+  }
+}
+
+extension Utf16Pointer on Pointer<Utf16> {
+  /// The number of UTF-16 code units in this zero-terminated UTF-16 string.
+  ///
+  /// The UTF-16 code units of the strings are the non-zero code units up to
+  /// the first zero code unit.
+  int get length {
+    final Pointer<Uint16> array = cast<Uint16>();
+    int length = 0;
+    while (array[length] != 0) {
+      length++;
+    }
+    return length;
+  }
+
+  /// Converts this UTF-8 encoded string to a Dart string.
+  ///
+  /// Decodes the UTF-16 code units of this zero-terminated code unit array as
+  /// Unicode code points and creates a Dart string containing those code
+  /// points.
+  ///
+  /// If [length] is provided, zero-termination is ignored and the result can
+  /// contain NUL characters.
+  String toDartString({int? length}) {
+    if (length != null) {
+      RangeError.checkNotNegative(length, 'length');
+    } else {
+      length = this.length;
```

Comment:
This seems inefficient. 

If you know the length, don't use a StringBuffer, just allocate a `Uint16List` of that length and copy the words directly into it, then call String.fromCharCodes (which has an optimized path for `Uint16List`).

When you don't know the length, you can use a `StringBuffer` and wait to see a NUL char, so you don't need an extra traversal to find the length.

So, I'd suggest two different functions, giving the compiler a good chance to inline and eliminate the first one based on the actual `length` argument (or its absence).
```dart
String toDartString({int? length}) => 
    length == null 
        ? _toUnknownLengthString(this.cast<Uint16>()) 
        : _toKnownLengthString(this.cast<Uint16>(), length);

static String _toKnownLengthString(Pointer<Uint16> words, int length) =>
    String.fromCharCodes(words.asTypedData(length));

String _toUnknownLengthString(Pointer<Uint16> words) {
  var buffer = StringBuffer();
  var i = 0;
  while (true) {
    var char = words.elementAt(i).value;
    if (char == 0) return buffer.toString();
    buffer.addCharCode(char);
  }
}
```